Upgrade to rails 4.2

allenwq opened the pull request:
Good thing is that most of the gems are compatible.

All tests passed in rails 4.2 and ruby 2.2


REF: http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html

wangqiang1208 commented on Gemfile:
```diff
@@ -5,19 +5,19 @@ source 'https://rails-assets.org'
 gem 'magiclabs-userstamp', github: 'lowjoel/userstamp'
 # Squeel as an SQL-like DSL
-gem 'squeel'
+gem 'squeel', github: 'danielrhodes/squeel'
```
Squeel is not working with rails 4.2, see https://github.com/activerecord-hackery/squeel/issues/352

lowjoel replied:
Wait for squeal then.

wangqiang1208 replied:
Few guys have fixed the issue in their own repo, the one I used is from one of them.

wangqiang1208 commented on bin/bundle:
```diff
@@ -1,3 +1,3 @@
-#!/usr/bin/env ruby2.1
```
Here are default configs generated by rails, Do we need to change to ruby2.2 ?

jsyeo replied:
> Do we need to change to ruby2.2 ? 

I think should just use simply `ruby`.

lowjoel replied:
you'll then be at the mercy of whichever ruby the user has installed... do we want that?

jsyeo replied:
Hmm. I'm not sure if this is the right place to ensure that people use the right ruby. :confused:

jsyeo replied:
We can use the ruby directive in the Gemfile to do this. http://bundler.io/v1.3/gemfile_ruby.html

lowjoel replied:
It was ruby2.1 because I was using the brightbox ruby-ng binaries, so that's what it gave me.

But I really don't know.

wangqiang1208 replied:
I actually prefer just use <code>ruby</code>.. cause then we no need to maintain here once we updated the ruby version.

And If we restrict to specific version, without that version user cannot run right ?

wangqiang1208 commented on spec/rails_helper.rb:
```diff
@@ -50,7 +50,6 @@
 
   # Ensure that all database seeds are in the database.
   config.before(:suite) do
-    Application.load_tasks
```
Don't have <code>load_tasks</code> anymore

lowjoel commented:
Let's wait for squeel and schema_plus to be fixed first. ActiveRecord now has foreign key support, so we should wait for them.

wangqiang1208 commented:
@lowjoel  Squeel has updated, how about schema_plus side ?

lowjoel commented:
schema_plus almost ready, I think.

wangqiang1208 commented:
Ok, then I'll do this again after we cleared some PRs.

lowjoel commented:
yah, I want to use my lowjoel/schema_plus_association_inverses@d5fdff3b2c gem :grinning:

wangqiang1208 commented:
Sure, that one will in a separate PR I guess ?

lowjoel commented:
Yah, but we first need to migrate to 4.2 first.

allenwq commented:
@lowjoel Here I've updated again.

The main changes are:

Upgrade to rails 4.2 and update our config/environments files

have_tag / have_form has deprecated from RSpec2 , removed them (tests start to fail in the new version of Spec, I think they completely removed these functions )

Update Gem schema_plus to their 2.0 version ( Have tried to re-run all the migrations, everything is fine)

allenwq commented on spec/features/global_announcements_spec.rb:
```diff
@@ -35,23 +34,21 @@
       end
 
       it 'shows the latest announcement' do
-        announcement = announcements.last
```
The <code>last</code> announcement cannot be found and I changed to <code>first </code>, something wrong with the implementation ?

allenwq commented on spec/features/global_announcements_spec.rb:
```diff
@@ -35,23 +34,21 @@
-        expect(page).to have_tag('div.global-announcement') do
-          have_tag('div.panel-heading', with_text: announcement.title)
-          have_tag('div.panel-body', with_text: announcement.content)
-        end
```
I think our old test was wrong. the block parsed to have_tag is not executed, it will always pass.
Things like this passed in my experiment:
```
        expect(page).to have_tag('div.global-announcement') do
          have_tag('looool,lol', with_text: announcement.title)
          have_tag('div.fasdf', with_text: announcement.content)
        end
```

allenwq commented:
@lowjoel  Turns out that I should include rspec machers in configs, don't know why it works previously without include it.

So I didn't modify any test cases in this PR, will fix the test cases which I mentioned in another PR

allenwq commented on config/application.rb:
```diff
@@ -6,19 +6,25 @@
-  # Settings in config/environments/* take precedence over those specified here.
-  # Application configuration should go into files in config/initializers
-  # -- all .rb files in that directory are automatically loaded.
+module Application
```
Can we permit both style ?

otherwise I'll need to define a module Application first :
```
module Application
end

class Application::Application
   .....
   ....
end

```

allenwq replied:
Forgot to say that the new configuration generated by rails 4.2 put the Application class inside a module. That's why I got this complain.

lowjoel replied:
I explicitly remove the application module because the rest of the classes we have are not in the application module...

allenwq replied:
Don't know you've removed that..then I'll just remove it.

allenwq commented:
@lowjoel done, ready for review :)

lowjoel commented:
What was broken?

allenwq commented:
I haven't found anything broken

testing using ruby 2.2.1

On Thursday, March 26, 2015, Joel Low <notifications@github.com> wrote:

> What was broken?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Coursemology/coursemology2/pull/50#issuecomment-86358275>
> .
>

lowjoel commented:
You said that the rspec html matcher doesn't accept blocks?

allenwq commented:
Yes, but that one not because rails 4.2. currently same, I said I will fix
it in another PR.

On Thu, Mar 26, 2015 at 2:27 PM, Joel Low <notifications@github.com> wrote:

> You said that the rspec html matcher doesn't accept blocks?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Coursemology/coursemology2/pull/50#issuecomment-86358955>
> .
>

lowjoel commented on config/environments/production.rb:
```diff
@@ -16,12 +16,13 @@
 
```
These changes are generated by the new Rails generator right?

allenwq replied:
Yes, like the new version is using <code>serve_static_files</code> instead of <code>serve_static_assets</code>

lowjoel commented on spec/rails_helper.rb:
```diff
@@ -53,6 +53,9 @@
   # https://relishapp.com/rspec/rspec-rails/docs
   config.infer_spec_type_from_file_location!
 
+  # Old school have_tag, with_tag(and more) matchers for rspec 3
```
...this needs set up? :open_mouth:

lowjoel replied:
okay, I was blur, sorry.

allenwq replied:
Yeah..Mentioned in their documents. I don't know why it works before.

lowjoel commented:
:+1:
